cli/pkg/commands/common/cmd.go

```diff
@@ -0,0 +1,106 @@
+/*
+Copyright 2024. Open Data Hub Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package common
+
+import (
+	"os"
+
+	tea "github.com/charmbracelet/bubbletea"
+	"github.com/opendatahub-io/ai-edge/cli/pkg/commands/flags"
+	"github.com/spf13/cobra"
+)
+
+// SubCommand is a type to represent the subcommand
+type SubCommand int
+
+const (
+	// SubCommandList is a subcommand to list items
+	SubCommandList SubCommand = iota
+	// SubCommandAdd is a subcommand to add items
+	SubCommandAdd
+)
+
+// NewCmd creates a new cobra command.
+//
+// The command will create a new tea program, passing the model created by the modelFactory, and run it.
+// The modelFactory will be called with the args, flags and subCommand.
+//
+// Example:
+// 	cmd := NewCmd(
+// 		"images",
+// 		"List images",
+// 		`List images`,
+// 		cobra.ExactArgs(3),
+// 		[]flags.Flag{flags.FlagModelRegistryUrl},
+// 		SubCommandList,
+// 		func(args []string, flags map[string]string, subCommand SubCommand) tea.Model {
+// 			return NewImagesModel(args, flags, subCommand)
+// 		},
+// 	)
+func NewCmd(
+	use, short, long string,
+	args cobra.PositionalArgs,
+	flags []flags.Flag,
+	command SubCommand,
+	modelFactory func(args []string, flags map[string]string, subCommand SubCommand) tea.Model,
+) *cobra.Command {
+
+	cmd := cobra.Command{
+		Use:   use,
+		Short: short,
+		Long:  long,
+		Args:  args,
+		Run: func(cmd *cobra.Command, args []string) {
+			ff := make(map[string]string)
+			for _, f := range flags {
+				v := ""
+				err := error(nil)
+				if f.IsInherited() {
+					v, err = cmd.InheritedFlags().GetString(f.String())
+					if err != nil {
+						cmd.PrintErrf("Error reading inherited flag %s: %v\n", f, err)
+						os.Exit(1)
+					}
+				} else {
+					v, err = cmd.Flags().GetString(f.String())
+					if err != nil {
+						cmd.PrintErrf("Error reading flag %s: %v\n", f, err)
+						os.Exit(1)
+					}
+				}
+				ff[f.String()] = v
+			}
+			_, err := tea.NewProgram(modelFactory(args, ff, command)).Run()
+			if err != nil {
+				cmd.PrintErrf("Error: %v\n", err)
+				os.Exit(1)
+			}
+		},
+	}
+
+	for _, f := range flags {
+		if !f.IsRootFlag() {
+			if f.IsInherited() {
+				cmd.PersistentFlags().StringP(f.String(), f.Shorthand(), f.Value(), f.Usage())
+			} else {
+				cmd.Flags().StringP(f.String(), f.Shorthand(), f.Value(), f.Usage())
+			}
+		}
+	}
```

Comment:
Here we're registering non-root flags only. Root flags are registered in `pkg/commands/root.go`. The `FlagsModelRegistryURL` is a root flag as declared in `pkg/commands/flags/flags.go`. 

One logical error that I just fixed, is that root flags should also be set as inherited. I pushed a fix. It wasn't producing a bug but it was logically incorrect. 

A root flag means it's defined at the root level, so can be inherited by all subcommand.

There can be a case of an inherited flag that's not root. In that case, it's defined by a specific sub-command and will be inherited by all of its nested sub-commands.

Reply:
@devguyio I see, I didn't realize that it is a root flag :)

However, in case of an inherited flag that is not root, if we add the third layer of commands, such a flag will be re-added at the third layer to the persistent flags, right? Because it will need to be listed among the flags to read and because it is the same slice, it will be iterated over when registering as well.
Right now it is not an issue as everything non-root should be registered. Also, maybe it won't be an issue even afterwards because cobra maybe can handle it.